odd-collector/odd_collector/adapters/snowflake/adapter.py

```diff
@@ -44,32 +65,41 @@ def get_data_entity_list(self) -> DataEntityList:
                     stage_type=raw_stage.stage_type,
                     downstream=raw_pipe.downstream,
                 )
-                for raw_stage in raw_stages
+                for raw_stage in self._raw_stages
                 if raw_pipe.stage_full_name == raw_stage.stage_full_name
             )
-        pipes_entities = [map_pipe(pipe, self.generator) for pipe in pipes]
-        tables = self._client.get_tables()
+        return [map_pipe(pipe, self.generator) for pipe in pipes]
 
-        tables_with_data_entities: List[
-            Tuple[Table, DataEntity]
-        ] = self._get_tables_entities(tables)
+    @cached_property
+    def _table_entities(self) -> list[tuple[Table, DataEntity]]:
+        return self._get_tables_entities(self._tables)
 
-        try:
-            tables_entities = [
-                table_with_entity[1] for table_with_entity in tables_with_data_entities
-            ]
-            schemas_entities = self._get_schemas_entities(tables_with_data_entities)
-            database_entity: DataEntity = self._get_database_entity(schemas_entities)
-
-            all_entities = [
-                *tables_entities,
-                *schemas_entities,
-                database_entity,
-                *pipes_entities,
-            ]
+    @cached_property
+    def _relationship_entities(self) -> list[DataEntity]:
+        return DataEntityRelationshipsMapper(
+            oddrn_generator=self.generator,
+            table_entities_pair=self._table_entities,
+        ).map(self._fk_constraints)
+
+    @cached_property
+    def _schema_entities(self) -> list[DataEntity]:
+        return self._get_schemas_entities(self._table_entities)
 
+    @cached_property
+    def _database_entity(self) -> DataEntity:
+        return self._get_database_entity(self._schema_entities)
```

Comment:
I removed redundant methods `_get_tables_entities()`, `_get_schemas_entities()` and `_get_database_entity()`, as I was just over encapsulated their logic and didn't pay attention to this.
So remaining `_table_entities()`, `_schema_entities()` and `_database_entity()` now have directly the mapping logic. I also think that we can leave `_schema_entities()` and `_database_entity()`, because `_database_entity()` is dependent on the `_schema_entities()` result.